File: Algorithm/Leetcode746.cpp
/* Leetcode 746 Min Cost Climbing Stairs */
/**
* On a staircase, the i-th step has some non-negative cost cost[i] assigned (0 indexed).
* Once you pay the cost, you can either climb one or two steps. You need to find minimum cost to reach the top of the floor, 
* and you can either start from the step with index 0, or the step with index 1.
*/
//Note:	1.cost will have a length in the range [2, 1000].
//		2.Every cost[i] will be an integer in the range [0, 999].

#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
	int minCostClimbingStairs(vector<int>& cost) {
		int dp[1001];				//dp[i]表示以i为终点所需的最小cost
		int n = cost.size();
		dp[0] = cost[0];
		dp[1] = cost[1];
		for (int i = 2; i < n; i++) {
			dp[i] = cost[i] + min(dp[i - 1], dp[i - 2]);
		}
		return min(dp[n - 1], dp[n - 2]);
	}
};

File: Algorithm/DP/LeetCode746.java
public class LeetCode746 {
    /* LeetCode 746 使用最小花费爬楼梯 */
    /**
     * 数组的每个索引作为一个阶梯，第 i个阶梯对应着一个非负数的体力花费值 cost[i](索引从0开始)。
     * 每当你爬上一个阶梯你都要花费对应的体力花费值，然后你可以选择继续爬一个阶梯或者爬两个阶梯。
     * 您需要找到达到楼层顶部的最低花费。在开始时，你可以选择从索引为 0 或 1 的元素作为初始阶梯。
     * 示例 1:
     * 输入: cost = [10, 15, 20]
     * 输出: 15
     * 解释: 最低花费是从cost[1]开始，然后走两步即可到阶梯顶，一共花费15。
     * 注意：
     * 1. cost 的长度将会在 [2, 1000]。
     * 2. 每一个 cost[i] 将会是一个Integer类型，范围为 [0, 999]。
     */
    public static void main(String[] args) {
        Solution1 solution1 = new Solution1();
        int ans = solution1.minCostClimbingStairs(new int[]{0, 1, 1, 0});
        System.out.println(ans);
        Solution solution = new Solution();
    }
}

// 优化：动态优化（滚动数组）
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        int first = cost[0];
        int second = cost[1];
        for(int i = 2; i < n; i++) {
            int curr = Math.min(first, second) + cost[i];
            first = second;
            second = curr;
        }
        return Math.min(first, second);
    }
}

// 解法：动态规划
class Solution1 {
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        int[] dp = new int[n];
        dp[0] = cost[0];
        dp[1] = cost[1];
        for(int i = 2; i < n; i++) {
            dp[i] = Math.min(dp[i - 2], dp[i - 1]) + cost[i];
        }
        return Math.min(dp[n - 2], dp[n - 1]);
    }
}
